File: client/typescript/src/result.ts
/* export interface Ok<T> {
    ok: true
    value: T
}

export interface Err<E> {
    ok: false
    value: E
}

export type Result<T, E> = Ok<T> | Err<E> */

type Ok<T> = { ok: T };
type Err<T> = { err: T };
type Result<T, E> = Ok<T> | Err<E>;

const foo: Result<number, string> = {
  ok: 123,
  err: "error",
};

File: src/reproca/result.py
from typing import Literal, Never

import msgspec


class UnwrapError(Exception):
    pass


class Ok[T](msgspec.Struct):
    value: T
    ok: Literal[True] = True

    def unwrap(self) -> T:
        return self.value


class Err[T](msgspec.Struct):
    value: T
    ok: Literal[False] = False

    def unwrap(self) -> Never:
        raise UnwrapError(self.value)


type Result[T, E] = Ok[T] | Err[E]
